[FSDP][Collectives] FSDP collectives skips communication ops and synchronizations when world size is 1

anshul-si opened the pull request:
**Summary**
In its current state, FSDP collectives uses cuda synchronizations and communication ops regardless of what the world size is. However, now that replicate will use FSDP, there will be instances where group size = 1 and these synchronizations and ops will be used needlessly. I have updated the collectives so that if group size = 1, these will be skipped. I have created a test case that uses CommDebugMode to ensure there are no communication ops. In addition, I set breakpoints in order to ensure that no streams were added to the device handle. 

https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/anshulsi_373eb66f-2460-4f90-b89c-6404dc7bd005_devgpu263.prn2.facebook.com_2773614.1753675022783198838.pt.trace.json

https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/anshulsi_a8e3f032-b06d-4d0f-a733-e8b7aef6e200_devgpu263.prn2.facebook.com_2768973.1753674993923532518.pt.trace.json

**Test Case**
pytest test/distributed/_composable/fsdp/test_fully_shard_training.py -k test_train_parity_single_worldsize1

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #159100



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta

sinhaanshul committed:
[FSDP][Collectives] FSDP collectives skips communication ops and synchronizations when world size is 1

[ghstack-poisoned]

weifengpy commented:
attach the trace showing copy-in and copy-out are indeed removed?

weifengpy commented on torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py:
```diff
@@ -356,6 +380,38 @@ def foreach_all_gather_copy_out(
     ) = all_gather_result
     _dtype, device = all_gather_output.dtype, all_gather_output.device
+
+    if group.size() == 1:
```
the code feels too repetitive to line 532 and onward. any chance to reduce the duplicate code?

weifengpy commented on torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py:
```diff
@@ -243,6 +243,30 @@ def foreach_all_gather(
+    if world_size == 1:
+        # Store the original parameters directly in all_gather_output for direct access in copy_out
+
+        param_all_gather_inputs = _get_param_all_gather_inputs(fsdp_params)
```
same duplicate code

sinhaanshul committed:
Update on "[FSDP][Collectives] FSDP collectives skips communication ops and synchronizations when world size is 1"

**Summary**
In its current state, FSDP collectives uses cuda synchronizations and communication ops regardless of what the world size is. However, now that replicate will use FSDP, there will be instances where group size = 1 and these synchronizations and ops will be used needlessly. I have updated the collectives so that if group size = 1, these will be skipped. I have created a test case that uses CommDebugMode to ensure there are no communication ops. In addition, I set breakpoints in order to ensure that no streams were added to the device handle. 

**Test Case**
pytest test/distributed/_composable/fsdp/test_fully_shard_training.py -k test_train_parity_single_worldsize1




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

sinhaanshul committed:
Update on "[FSDP][Collectives] FSDP collectives skips communication ops and synchronizations when world size is 1"

**Summary**
In its current state, FSDP collectives uses cuda synchronizations and communication ops regardless of what the world size is. However, now that replicate will use FSDP, there will be instances where group size = 1 and these synchronizations and ops will be used needlessly. I have updated the collectives so that if group size = 1, these will be skipped. I have created a test case that uses CommDebugMode to ensure there are no communication ops. In addition, I set breakpoints in order to ensure that no streams were added to the device handle. 

**Test Case**
pytest test/distributed/_composable/fsdp/test_fully_shard_training.py -k test_train_parity_single_worldsize1




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

sinhaanshul committed:
Update on "[FSDP][Collectives] FSDP collectives skips communication ops and synchronizations when world size is 1"

**Summary**
In its current state, FSDP collectives uses cuda synchronizations and communication ops regardless of what the world size is. However, now that replicate will use FSDP, there will be instances where group size = 1 and these synchronizations and ops will be used needlessly. I have updated the collectives so that if group size = 1, these will be skipped. I have created a test case that uses CommDebugMode to ensure there are no communication ops. In addition, I set breakpoints in order to ensure that no streams were added to the device handle. 

**Test Case**
pytest test/distributed/_composable/fsdp/test_fully_shard_training.py -k test_train_parity_single_worldsize1




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

anshul-si commented:
Closing PR due to new approach
